src/Workspaces/Core/Portable/Rename/Renamer.RenameDocumentActionSet.cs

```diff
@@ -0,0 +1,120 @@
+
+            /// <summary>
+            /// Applies each <see cref="RenameDocumentAction"/> in order and returns the final solution. 
+            /// All actions must be contained in <see cref="ApplicableActions" />
+            /// </summary>
+            /// <remarks>
+            /// An empty action set is still allowed and will return a modified solution
+            /// that will update the document properties as appropriate.
+            /// </remarks>
+            public async Task<Solution> UpdateSolutionAsync(Solution solution, ImmutableArray<RenameDocumentAction> actions, CancellationToken cancellationToken)
+            {
+                if (actions.Any(a => !ApplicableActions.Contains(a)))
+                {
+                    throw new ArgumentException(WorkspacesResources.Cannot_apply_action_that_is_not_in_applicableactions);
+                }
+
+                // Prior to updating the solution it's possible the document id has changed between the time 
+                // the document action info was generated and when the solution update is applied. We
+                // do a best effort to still locate the document if the id has changed.
+                var document = GetDocument(solution);
+
+                // If the document was found in the solution then the current id will be durable across actions
+                // since we own the solution snapshot at this point. 
+                var documentId = document.Id;
+
+                // Make sure that the document name and folders are updated to what we expect them to be
+                solution = solution
+                    .WithDocumentName(documentId, _documentName)
+                    .WithDocumentFolders(documentId, _documentFolders);
+
+                // Apply each action individually. Order should not matter
+                foreach (var action in actions)
+                {
+                    document = solution.GetRequiredDocument(documentId);
+                    solution = await action.GetModifiedSolutionAsync(document, cancellationToken).ConfigureAwait(false);
+                }
+
+                return solution;
+            }
+
+            /// <summary>
+            /// Attempts to find the document in the solution. Tries by documentId first, but 
+            /// that's not always reliable between analysis and application of the rename actions
+            /// </summary>
+            private Document GetDocument(Solution solution)
+            {
+                // DocumentId is the best bet for finding a document,
+                // but it's possible the document was renamed or moved
+                // before actions were applied.
+                if (solution.ContainsDocument(_documentId))
+                {
+                    return solution.GetRequiredDocument(_documentId);
+                }
+
+                var project = solution.GetRequiredProject(_projectId);
```

Comment:
Does this need to throw a more useful exception if the caller passed in a bad Solution? Or is whatever this throws good? (It actually might be, I don't know.)

Reply:
It will throw "Project_of_ID_0_is_required_to_accomplish_the_task_but_is_not_available_from_the_solution" 

which... is enough? maybe?